earth_enterprise/src/fusion/autoingest/AssetFactory_unittest.cpp

```diff
@@ -0,0 +1,200 @@
+      // Leaving the subtype changeable by the tests via testSubTypeToUseForStringConstructor;
+      MockAssetStorage storage;
+      storage.name = ref_;
+      storage.type = MockAssetImpl::EXPECTED_TYPE;
+      storage.subtype = testSubTypeToUseForStringConstructor;
+      
+      MockAssetConfig config;
+      impl = std::make_shared<MockAssetImpl>(storage, config);
+    }
+
+    // For the `if (asset)` check in MakeNew
+    explicit operator bool(void) const
+    {
+      return impl != nullptr;
+    }
+
+    Impl* operator->(void) const {
+      return impl.get();
+    }
+};
+
+string MockMutableAsset::testSubTypeToUseForStringConstructor;
+
+class AssetFactoryTest : public testing::Test {
+ public:
+  AssetFactoryTest() 
+  {
+    MockAssetImpl::EXPECTED_TYPE = AssetDefs::Imagery;
+    MockAssetImpl::EXPECTED_SUBTYPE = "mockSubtype";
+    MockMutableAsset::testSubTypeToUseForStringConstructor = "someOtherSubtype";
+  }
+};
+
+// TEST DATA
+string testAssetRef = "/gevol/assets/AssetRef",
+       testAssetRef1 = "/gevol/assets/AssetRef1",
+       testAssetRef2 = "/gevol/assets/AssetRef2";
+std::vector<SharedString> testInputs { "Input1", "Input2"},
+                          testInputs1 { "Inputs3", "Inputs4", "Inputs5" };
+khMetaData testMeta;
+MockAssetConfig testConfig;
+
+TEST_F(AssetFactoryTest, MakeNew) {
+  MockMutableAsset handle = MakeNew<MockMutableAsset, MockAssetConfig>(
+      testAssetRef, testInputs, testMeta, testConfig);
+  ASSERT_EQ(handle.impl->name, testAssetRef);
+  ASSERT_EQ(handle.impl->inputs, testInputs);
+  ASSERT_EQ(handle.impl->meta, testMeta);
+  ASSERT_EQ(handle.impl->type, AssetDefs::Imagery);
+}
+
+// ASSERT_THROW seems to trip up on function templates with more than one
+// template parameter. Making a function pointer helps it get past that.
+MockMutableAsset (*pMakeNew)( const std::string &ref_, 
+                                    const std::vector<SharedString>& inputs_,
+                                    const khMetaData &meta, 
+                                    const MockAssetConfig &config) =
+  MakeNew<MockMutableAsset, MockAssetConfig>;
+TEST_F(AssetFactoryTest, MakeNewAlreadyExists) {
+  // Make sure the std::string constructor for MockMutableAsset will have the same
+  // subtype as the one we're trying to create in order to induce an exception.
+
+  MockMutableAsset::testSubTypeToUseForStringConstructor = MockAssetImpl::EXPECTED_SUBTYPE;
+  ASSERT_THROW(pMakeNew(testAssetRef, testInputs, testMeta, testConfig), khException);
+}
+
+TEST_F(AssetFactoryTest, FindMake_New)
+{
+    MockMutableAsset handle = FindMake<MockMutableAsset, MockAssetConfig>
+            (testAssetRef1, AssetDefs::Imagery, testInputs1, testMeta, testConfig);
+    ASSERT_EQ(handle.impl->name, testAssetRef1);
+    ASSERT_EQ(handle.impl->inputs, testInputs1);
+    ASSERT_EQ(handle.impl->meta, testMeta);
+    ASSERT_EQ(handle.impl->type, AssetDefs::Imagery);
+}
+
+TEST_F(AssetFactoryTest, FindMake_Exists)
```

Comment:
If I understand correctly, this isn't testing what is intended. Both calls to FindMake skip over the if statement in the middle of the FindMake function, so none of them is falling into the if statement.

If you want to be completely comprehensive, you've got 3 FindMake functions with 2 possible paths (go into the if statement or skip over it). You could write tests that hit all 6 possible paths.

Reply:
I'm still not finding any tests that go into the if statement in any of the FindMake functions. I would feel better about the tests if at least one of them went into the if statement.

Reply:
Ok, the latest push fixed my issue with the tests. You're now going into the if statement in all 3 versions of FindMake.

If you want to improve the tests further you could update your `FindMake_New` to also call all 3 versions of FindMake but not go into the if statement on any of them. However, that's more of icing on the cake; I'm OK with it as is.